btcec/schnorr/musig2/context.go

```diff
@@ -0,0 +1,280 @@
+	signingKey *btcec.PrivateKey
+
+	// pubKey is our even-y coordinate public  key.
+	pubKey *btcec.PublicKey
+
+	// keySet is the set of all signers.
+	keySet []*btcec.PublicKey
+
+	// combinedKey is the aggregated public key.
+	combinedKey *btcec.PublicKey
+
+	// uniqueKeyIndex is the index of the second unique key in the keySet.
+	// This is used to speed up signing and verification computations.
+	uniqueKeyIndex int
+
+	// keysHash is the hash of all the keys as defined in musgi2.
+	keysHash []byte
+
+	// shouldSort keeps track of if the public keys should be sorted before
+	// any operations.
+	shouldSort bool
+}
+
+// NewContext creates a new signing context with the passed singing key and set
+// of public keys for each of the other signers.
+//
+// NOTE: This struct should be used over the raw Sign API whenever possible.
+func NewContext(signingKey *btcec.PrivateKey,
+	signers []*btcec.PublicKey, shouldSort bool) (*Context, error) {
+
+	// As a sanity check, make sure the signing key is actually amongst the sit
+	// of signers.
+	//
+	// TODO(roasbeef): instead have pass all the _other_ signers?
+	pubKey, err := schnorr.ParsePubKey(
+		schnorr.SerializePubKey(signingKey.PubKey()),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyFound bool
+	for _, key := range signers {
+		if key.IsEqual(pubKey) {
+			keyFound = true
+			break
+		}
+	}
+	if !keyFound {
+		return nil, ErrSignerNotInKeySet
+	}
+
+	// Now that we know that we're actually a signer, we'll generate the
+	// key hash finger print and second unique key index so we can speed up
+	// signing later.
+	keysHash := keyHashFingerprint(signers, shouldSort)
+	uniqueKeyIndex := secondUniqueKeyIndex(signers, shouldSort)
+
+	// Next, we'll use this information to compute the aggregated public
+	// key that'll be used for signing in practice.
+	combinedKey := AggregateKeys(
+		signers, shouldSort, WithKeysHash(keysHash),
+		WithUniqueKeyIndex(uniqueKeyIndex),
+	)
+
+	return &Context{
+		signingKey:     signingKey,
+		pubKey:         pubKey,
+		keySet:         signers,
+		combinedKey:    combinedKey,
+		uniqueKeyIndex: uniqueKeyIndex,
+		keysHash:       keysHash,
+		shouldSort:     shouldSort,
+	}, nil
+}
+
+// CombinedKey returns the combined public key that will be used to generate
+// multi-signatures  against.
+func (c *Context) CombinedKey() btcec.PublicKey {
+	return *c.combinedKey
+}
+
+// PubKey returns the public key of the signer of this session.
+func (c *Context) PubKey() btcec.PublicKey {
+	return *c.pubKey
+}
+
+// Session represents a musig2 signing session. A new instance should be
+// created each time a multi-signature is needed. The session struct handles
+// nonces management, incremental partial sig vitrifaction, as well as final
+// signature combination. Errors are returned when unsafe behavior such as
+// nonce re-use is attempted.
+//
+// NOTE: This struct should be used over the raw Sign API whenever possible.
+type Session struct {
```

Comment:
I really like this API! It should make this quite easy to use and also safe.

One use case I was thinking about that wouldn't be possible with the current API is the channel funding 2-of-2 MuSig2 case. Because if we want to send our nonces over in the `open_channel` message, we wouldn't know the other party's public key yet, so couldn't create a full context yet. But we'd need a session to get access to "safe" nonces.

Maybe we can combine the two structs and keep some kind of progression state for the different steps?

Reply:
> One use case I was thinking about that wouldn't be possible with the current API is the channel funding 2-of-2 MuSig2 case. Because if we want to send our nonces over in the open_channel message, we wouldn't know the other party's public key yet,

Great point! I was coming at this from the PoV of initial commitment signing, where we already had created the joint public key at that point... 

One way to handle this would be to add a functional argument here that can be passed in which just uses the nonces specified by the caller. This would also be useful for deterministic tests. This way, you either initialize with the set of public keys (then the nonce is produced), or you do it with just your key (in order to gain access to the nonce) with you specifying the keys of the remote party in a future step. 

Would need to think about this a bit though, as it then exposes the caller (somewhat) to nonce generation themselves, but I think we can wrap that up at a high level.

Reply:
Or we permit passing in only a single key there, generate the nonces as normal, but return errors from certain calls until the other key is registered? So we'd allow a dynamic key set, with a similar `Register` method for the additional set of keys.

Reply:
Also re context vs session, my initial idea was that both sides would maintain the some _context_, and then make new sessions on the fly each time they need to piggy back nonces into messages. By next week or so, I should have a more concrete sketch of the flow here both during funding as well as normal commitment updates.